ocaml/xenopsd/lib/topology.ml

```diff
@@ -194,25 +187,19 @@ module NUMA = struct
                None
          )
     in
-    let single_nodes =
-      valid_nodes
-      |> Seq.map (fun i ->
-             let self_distance = d.(i).(i) in
-             (distance_to_candidate self_distance, Seq.return i)
-         )
-    in
     let numa_nodes = Array.length d in
     let nodes =
       if numa_nodes > 16 then
-        (* try just the single nodes, and give up (use all nodes otherwise) to
-           avoid exponential running time. We could do better here, e.g. by
+        (* Avoid generating too many candidates because of the exponential
+           running time. We could do better here, e.g. by
            reducing the matrix *)
-        single_nodes
-      else
         valid_nodes
-        |> seq_all_subsets
-        |> Seq.filter_map (node_distances d)
-        |> seq_append single_nodes
```

Comment:
I think the single nodes were always appended just in case something goes wrong with the filtering algorithm (which was meant to be somewhat smarter than brute force, or it may evolve in the future to be somewhat smarter).

As the algorithm currently looks like I agree that we don't need to append the single nodes here. Perhaps this would be a good condition to test for in the testcases, that single (reachable) nodes are always present with the expected value (unless such a test already exists).